### PR TITLE
In-hand omnitools auto-switch when a deconstruction tool contextAction is selected

### LINKED ISSUE #12305: Desctruction device for borgs
### Labels

[BUG][MINOR]

### Description

While using the context action of the omnitool , it takes out the context action of the deconstruction device , making it harder for borgs to destroy stuff

### Steps to reproduce

1. Borg time
2. Destroy anything
3. Take 5 hours because of the context action disabling each other
4. Cry

### Expected Behavior

both context actions appear or omnitool context action dont disable the destroy context action , enabling it again once the omnitool context action is out

### Additional Information & Screenshots

Reported by: RubiCubie
On server: Goonstation 1 Classic: Heisenbee
Round log date: 2022-12-05-14-45
Reported on: 2022-12-05 15:03:40
Map: DONUT3

### PR DESCRIPTION
**Pull request by pgmzeta**
## About the PR 
When holding an omnitool and a deconstruction device, the omnitool will automatically switch to the matching required tool (if the omnitool has it), then automatically apply the tool after 0.5 seconds.

* Adds robot-specific handling for checking if an item is equipped
  * Cursory review showed nothing that would break with this change
* Moves omnitool mode defines into globals
* Restructures the deconstruction context menu code
  * Subclasses define omnitool mode/path and success text/sound
 
## Why's this needed? 
Allows for graceful use of omnitools with deconstruction devices
Fixes #12305

## Changelog 

```changelog
(u)glowbold
(*)Omnitools auto-swap to the needed tool when using the deconstruction tool context menu.
```

### CODE REVIEW

**Comment by thegrb93** [comment]
Wouldn't this be bypassing the required knowledge to use the tool? I thought having the tool is half of what's required, and knowing how to use it is the other half.

**Comment by CheffieGithub** [comment]
I think omni defines sould be in _std now they are in two files? minus the wait time one.

File: code/modules/interface/multiContext/context_actions.dm [hunk 1]
```diff
@@ -480,6 +489,16 @@
...
+				if(istype(I, /obj/item/tool/omnitool))
+					var/obj/item/tool/omnitool/omni = I
+					if (!(OMNI_MODE_WRENCHING in omni.modes))
+						continue
+					omni.change_mode(OMNI_MODE_WRENCHING, user, /obj/item/wrench)
+					user.show_text("You flip [omni] to wrenching mode...", "blue")
+					sleep(OMNI_TOOL_WAIT_TIME)
+					user.show_text("...then wrench [target]'s bolts.", "blue")
+					playsound(target, 'sound/items/Crowbar.ogg', 50, 1)
+					return ..()
```
Comment (TobleroneSwordfish) [inline code/modules/interface/multiContext/context_actions.dm:501]: could move this to the parent and define an `omni_mode` var on each child?

File: code/modules/interface/multiContext/context_actions.dm [hunk 2]
```diff
@@ -442,6 +442,15 @@
...
 			. = ..()
 
+// bodge to support omnitools
+#define OMNI_MODE_PRYING 1
+#define OMNI_MODE_SNIPPING 2
+#define OMNI_MODE_WRENCHING 3
+#define OMNI_MODE_SCREWING 4
+#define OMNI_MODE_PULSING 5
+#define OMNI_MODE_CUTTING 6
+#define OMNI_MODE_WELDING 7
```
Comment (TobleroneSwordfish) [inline code/modules/interface/multiContext/context_actions.dm:452]: maybe better to just make these global rather than redefining them here?

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]

File: code/modules/interface/multiContext/context_actions.dm (hunk 2 above)
Comment (pgmzeta) [reply code/modules/interface/multiContext/context_actions.dm]: [dc72de6](https://github.com/goonstation/goonstation/pull/12356/commits/dc72de6950478e20c743160c28bba7d5b31181a7)

**Comment by M-Earthfire** [comment]
> Wouldn't this be bypassing the required knowledge to deconstruct? I thought having the tool is half of what's required, and knowing how to do it is the other half.

The deconstruction tool does litterally say what it needs when you click an object. So you just need the proper tools at hand, which the omnitool already have build-in.

File: code/modules/interface/multiContext/context_actions.dm [hunk 3]
```diff
@@ -611,15 +603,6 @@
...
 					return ..()
 
-#undef OMNI_MODE_PRYING
-#undef OMNI_MODE_SNIPPING
-#undef OMNI_MODE_WRENCHING
-#undef OMNI_MODE_SCREWING
-#undef OMNI_MODE_PULSING
-#undef OMNI_MODE_CUTTING
-#undef OMNI_MODE_WELDING
-#undef OMNI_TOOL_WAIT_TIME
```
Comment (CheffieGithub) [inline code/modules/interface/multiContext/context_actions.dm]: You still still be undefining wait_time no?

**Comment by thegrb93** [comment]
> > Wouldn't this be bypassing the required knowledge to deconstruct? I thought having the tool is half of what's required, and knowing how to do it is the other half.
> 
> The deconstruction tool does litterally say what it needs when you click an object. So you just need the proper tools at hand, which the omnitool already have build-in.

Knowing that info beforehand is the difference between being inefficient and efficient imo. Seems like a pretty decent buff to silicon's already advantageous omnitool.

**Comment by CheffieGithub** [comment]
> Knowing that info beforehand is the difference between being inefficient and efficient imo. Seems like a pretty decent buff to silicon's already advantageous omnitool.

I think almost every player knows how to decon quickly, its really not rocket science

**Comment by pgmzeta** [comment]
> Wouldn't this be bypassing the required knowledge to deconstruct? I thought having the tool is half of what's required, and knowing how to do it is the other half.

Using a deconstruction device on machinery will show you a context menu with the exact tool type needed to dismantle it back into a frame. This wouldn't be used for e.g. dismantling a wall where the steps are not evident from the get-go.

File: code/modules/interface/multiContext/context_actions.dm (hunk 1 above)
Comment (pgmzeta) [reply code/modules/interface/multiContext/context_actions.dm]: restructured to include this in [3eafef4](https://github.com/goonstation/goonstation/pull/12356/commits/3eafef46ef91b737d02a9a0c1a508f6f71f4ab69)

File: code/modules/interface/multiContext/context_actions.dm (hunk 3 above)
Comment (pgmzeta) [reply code/modules/interface/multiContext/context_actions.dm]: [a36418c](https://github.com/goonstation/goonstation/pull/12356/commits/a36418ce24fd8cac28c02e94d81bc9d5670f8edb)

**Comment by thegrb93** [comment]
> > Knowing that info beforehand is the difference between being inefficient and efficient imo. Seems like a pretty decent buff to silicon's already advantageous omnitool.
> 
> I think almost every player knows how to decon quickly, its really not rocket science

So you're saying this is unneeded then

**Comment by CheffieGithub** [comment]
> So you're saying this is unneeded then

No im not, its needed because the two context menus clash and make it very slow to decon things. 
My point is that it being automatic doesn't change anything since its very easy to know how to decon something and anyone who has deconned something before can quickly do it for anything else.

**Comment by pgmzeta** [comment]
> So you're saying this is unneeded then

The root of the issue is that machinery deconstruction can only be done with the deconstruction tool context actions menu shown - e.g. the little icons for weld/pulse/etc. When switching which tool type to use with omnitools, it brings up a separate context action menu that closes the deconstruction tool menu. The end result is that currently you have to re-open both the omnitool and deconstrction device context action menus for every deconstruction step when using an omnitool.

This does make it significantly easier to use the omnitool when using a deconstruction device than the pre-change omnitool, but I'm unsure of a better approach to solve the current usability problem. I've added a small delay between steps when using an omnitool to allow for the cyborg omnitool animation to finish and to help balance the usage.

I will note that cyborg omnitools do not have a weldingtool mode, so some deconstruction tool actions will require manually switching to the welding tool and clicking.

**Comment by frawhst** [comment]
Feels fine

**Pull request merged**